cassandane/Cassandane/Cyrus/JMAPMailbox.pm

```diff
@@ -5282,4 +5296,64 @@ sub test_mailbox_ignore_notes_subfolders
     $self->assert_str_equals($inboxId, $res->[1][1]{list}[0]{id});
 }
 
+sub test_mailbox_set_create_specialuse_nochildren
+    :min_version_3_7 :needs_component_jmap :NoStartInstances
+{
+    my ($self) = @_;
+
+    $self->{instance}->{config}->set('specialuse_nochildren' => '\\Trash');
+    $self->_start_instances();
+    $self->_setup_http_service_objects();
+    $self->setup_default_using();
+
+    # set up a Trash folder with \Trash special-use annotation
+    my $imaptalk = $self->{store}->get_client();
+    $imaptalk->create("Trash", "(USE (\\Trash))");
+    $self->assert_equals('ok', $imaptalk->get_last_completion_response());
+
+    my $mailboxid = Cyrus::DList->parse_string(
+                        $imaptalk->get_response_code('mailboxid')
+                    );
+    my ($trash_id) = @{$mailboxid->as_perl()};
+
+    # okay lets go
+    my $jmap = $self->{jmap};
+
+    # should not be able to create a child of \Trash
+    my $res = $jmap->CallMethods([['Mailbox/set', {
+        create => {
+            1 => {
+                parentId => $trash_id,
+                name => 'child',
+            },
+        },
+    }, "R1"]]);
+
+    # XXX that syslogs an IOERROR, surprisingly -- ignore it
+    $self->{instance}->getsyslog();
+
+    $self->assert_null($res->[0][1]->{created});
+    $self->assert_deep_equals({ 1 => { type => 'forbidden' } },
+                              $res->[0][1]->{notCreated});
+
+    # what if we remove the annotation
+    $imaptalk->setmetadata("Trash", "/private/specialuse", undef);
```

Comment:
This could be a `Mailbox/set{update}` where you set `role => undef`

Reply:
It looks like this might not be allowed:

```
      'notUpdated' => {
        '21c5642b-63b7-4bd3-8149-22aea5229f23' => {
          'description' => 'The final mailbox state is invalid due to role conflicts:
A mailbox had protected specialuse \\Trash but this specialuse is missing in the new mailbox state.',
          'properties' => [
            'role',
          ],
          'type' => 'invalidProperties',
        },
      },
```

Do I need to do something special to remove the trash role?

Reply:
Now that I think of it: that's because the `specialuse_protect` imapd.conf option does not allow to delete `Trash`. I am surprised this works over IMAP.

Reply:
Ohhh that's interesting.

Reply:
We talked about this and decided to just use IMAP for this stage after all, because working around the more restrictive JMAP behaviour would just add complexity that wasn't relevant to the behaviour being tested.